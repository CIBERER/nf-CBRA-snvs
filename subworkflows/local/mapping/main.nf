include { BWA_MEM                 } from '../../../modules/nf-core/bwa/mem/main'
include { GATK4_FASTQTOSAM } from '../../../modules/nf-core/gatk4/fastqtosam/main'
include { GATK4_MERGEBAMALIGNMENT } from '../../../modules/nf-core/gatk4/mergebamalignment/main'
include { GATK4SPARK_MARKDUPLICATES } from '../../../modules/nf-core/gatk4spark/markduplicates/main'
include { PICARD_SORTSAM } from '../../../modules/nf-core/picard/sortsam/main'
include { GATK4_BASERECALIBRATOR } from '../../../modules/nf-core/gatk4/baserecalibrator/main'
include { GATK4_APPLYBQSR } from '../../../modules/nf-core/gatk4/applybqsr/main'
include { SAMTOOLS_SORT   } from '../../../modules/nf-core/samtools/sort/main'
include { SAMTOOLS_INDEX   } from '../../../modules/nf-core/samtools/index/main'


workflow MAPPING {

    take:
    ch_reads        // channel (mandatory): [ val(meta), [ path(reads) ] ]
    ch_index        // channel (mandatory): [ val(meta2), path(index) ]
    ch_fasta        // channel (mandatory) : [ val(meta2), path(fasta) ]
    ch_fai          // channel (mandatory) : [ val(meta2), path(fai) ]
    ch_refdict      // channel (mandatory) : [ val(meta2), path(dict) ]
    ch_intervals    // channel (mandatory) : [ val(meta), path(bed) ]
    ch_known_sites  // channel (mandatory) : [ val(meta3), path(vcf) ]
    ch_known_sites_tbi  // channel (mandatory) : [ val(meta3), path(vcf) ]

    main:

    ch_versions = Channel.empty()
    
    BWA_MEM (
        ch_reads,
        ch_index,
        ch_fasta,
        false
    )
    ch_versions = ch_versions.mix(BWA_MEM.out.versions.first())

    GATK4_FASTQTOSAM (
        ch_reads
    )
    ch_versions = ch_versions.mix(GATK4_FASTQTOSAM.out.versions.first())

    GATK4_MERGEBAMALIGNMENT (
        BWA_MEM.out.bam.join(GATK4_FASTQTOSAM.out.bam),
        ch_fasta,
        ch_refdict
    )
    ch_versions = ch_versions.mix(GATK4_MERGEBAMALIGNMENT.out.versions.first())

    GATK4SPARK_MARKDUPLICATES (
        GATK4_MERGEBAMALIGNMENT.out.bam,
        ch_fasta.map {meta , fasta -> [fasta]},
        ch_fai.map {meta, fai -> [fai]}, 
        ch_refdict.map {meta, dict -> [dict]}
    )
    ch_versions = ch_versions.mix(GATK4SPARK_MARKDUPLICATES.out.versions.first())

    GATK4_BASERECALIBRATOR (   
         GATK4SPARK_MARKDUPLICATES.out.output.join(GATK4SPARK_MARKDUPLICATES.out.bam_index.join(ch_intervals)),
         ch_fasta.map {meta , fasta -> [fasta]},
         ch_fai.map {meta, fai -> [fai]}, 
         ch_refdict.map {meta, dict -> [dict]},
         ch_known_sites,
         ch_known_sites_tbi
    )
    ch_versions = ch_versions.mix(GATK4_BASERECALIBRATOR.out.versions.first())

    GATK4_APPLYBQSR (
        GATK4SPARK_MARKDUPLICATES.out.output.join(GATK4SPARK_MARKDUPLICATES.out.bam_index).join(GATK4_BASERECALIBRATOR.out.table).join(ch_intervals),
        ch_fasta.map {meta , fasta -> [fasta]},
        ch_fai.map {meta, fai -> [fai]}, 
        ch_refdict.map {meta, dict -> [dict]}
    )
    ch_versions = ch_versions.mix(GATK4_APPLYBQSR.out.versions.first())

    SAMTOOLS_INDEX (
        GATK4_APPLYBQSR.out.bam 
    )
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

    emit:
    bam = GATK4_APPLYBQSR.out.bam.join(SAMTOOLS_INDEX.out.bai) // channel: [ [val(meta)], path(bam), path(bai)]
    versions = ch_versions                     // channel: [ versions.yml ]

}