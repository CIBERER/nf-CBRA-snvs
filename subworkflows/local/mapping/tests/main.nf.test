nextflow_workflow {

    name "Test Subworkflow MAPPING"
    script "../main.nf"
    workflow "MAPPING"

    tag "subworkflows"
    tag "subworkflows/mapping"
    tag "bwa"
    tag "bwa/mem"
    tag "gatk4"
    tag "gatk4/fastqtosam"
    tag "gatk4/mergebamalignment"
    tag "gatk4/fastqtosam"
    tag "gatk4/baserecalibrator"
    tag "gatk4/applybqr"
    tag "gatk4spark"
    tag "gatk4spark/markduplicates"    
    tag "picard"
    tag "picard/sortsam"

    test("mapping - fastq - with intervals") {

        setup {

            run("BWA_INDEX") {
                script "../../../../modules/nf-core/bwa/index/main.nf"
                process {
                    """
                    input[0] = [
                        [ id: 'test'],
                        file("https://zenodo.org/records/11119470/files/chr17_var_regions.fa", checkIfExists: true),
                    ]
                    """
                }
            }
            run("PICARD_CREATESEQUENCEDICTIONARY") {
                script "../../../../modules/nf-core/picard/createsequencedictionary/main.nf"
                process {
                    """
                    input[0] = [
                        [ id: 'chr17_var_regions'],
                        file("https://zenodo.org/records/11119470/files/chr17_var_regions.fa", checkIfExists: true),
                    ]
                    """
                }
            }
        }

        when {
            workflow {
                """
                input[0] = [
                    [ id : 'Sample_zenodo' ],
                    file("https://zenodo.org/records/13838219/files/my_R1.fastq", checkIfExists: true),
                    file("https://zenodo.org/records/13838219/files/my_R2.fastq", checkIfExists: true)
                ]
                input[1] = BWA_INDEX.out.index
                input[2] =  [
                        [ id: 'test'],
                        file("https://zenodo.org/records/11119470/files/chr17_var_regions.fa", checkIfExists: true)
                    ]
                input[3] = file("https://zenodo.org/records/11119470/files/chr17_var_regions.fa.fai", checkIfExists: true)
                input[4] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                input[5] = file("https://zenodo.org/records/11185276/files/my.bed", checkIfExists: true)
                input[6] = []
                input[7] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
