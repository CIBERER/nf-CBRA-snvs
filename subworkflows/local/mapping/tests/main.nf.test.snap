{
    "mapping - fastq - with intervals": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam:md5,23afbb8349bdafa4f6216913ae20714a",
                        "test_sample.bam.bai:md5,319dc3f628ca5e0617189c52ece9bfcc"
                    ]
                ],
                "1": [
                    "versions.yml:md5,080f0e7ce02ed9df7ae0b9911f12b015",
                    "versions.yml:md5,416c404b2a72368d4108f89497a35eec",
                    "versions.yml:md5,4865167bc121901e9a2bf78a36bb79ee",
                    "versions.yml:md5,4f84c70c49f1c407406f23390d7900ab",
                    "versions.yml:md5,a70ba4c1556659edb743f9748138e7ab",
                    "versions.yml:md5,b92899c7abb47fa1755de7145e7100e8",
                    "versions.yml:md5,ca649b5af272fa3a9ae828fb35e9582e"
                ],
                "bam": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam:md5,23afbb8349bdafa4f6216913ae20714a",
                        "test_sample.bam.bai:md5,319dc3f628ca5e0617189c52ece9bfcc"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,080f0e7ce02ed9df7ae0b9911f12b015",
                    "versions.yml:md5,416c404b2a72368d4108f89497a35eec",
                    "versions.yml:md5,4865167bc121901e9a2bf78a36bb79ee",
                    "versions.yml:md5,4f84c70c49f1c407406f23390d7900ab",
                    "versions.yml:md5,a70ba4c1556659edb743f9748138e7ab",
                    "versions.yml:md5,b92899c7abb47fa1755de7145e7100e8",
                    "versions.yml:md5,ca649b5af272fa3a9ae828fb35e9582e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-01-15T08:20:17.707172247"
    },
    "mapping - fastq - without intervals": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam:md5,00afa2d38996dfee4cc66b274445d4f7",
                        "test_sample.bam.bai:md5,bb9ad63f7a664ceb2f5c730facad75e5"
                    ]
                ],
                "1": [
                    "versions.yml:md5,080f0e7ce02ed9df7ae0b9911f12b015",
                    "versions.yml:md5,416c404b2a72368d4108f89497a35eec",
                    "versions.yml:md5,4865167bc121901e9a2bf78a36bb79ee",
                    "versions.yml:md5,4f84c70c49f1c407406f23390d7900ab",
                    "versions.yml:md5,a70ba4c1556659edb743f9748138e7ab",
                    "versions.yml:md5,b92899c7abb47fa1755de7145e7100e8",
                    "versions.yml:md5,ca649b5af272fa3a9ae828fb35e9582e"
                ],
                "bam": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam:md5,00afa2d38996dfee4cc66b274445d4f7",
                        "test_sample.bam.bai:md5,bb9ad63f7a664ceb2f5c730facad75e5"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,080f0e7ce02ed9df7ae0b9911f12b015",
                    "versions.yml:md5,416c404b2a72368d4108f89497a35eec",
                    "versions.yml:md5,4865167bc121901e9a2bf78a36bb79ee",
                    "versions.yml:md5,4f84c70c49f1c407406f23390d7900ab",
                    "versions.yml:md5,a70ba4c1556659edb743f9748138e7ab",
                    "versions.yml:md5,b92899c7abb47fa1755de7145e7100e8",
                    "versions.yml:md5,ca649b5af272fa3a9ae828fb35e9582e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-01-15T08:21:02.306854574"
    }
}